geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/LayerMetadataStore.java

```diff
@@ -0,0 +1,401 @@
+     *     java.lang.String)
+     */
+    public void putEntry(final String layerName, final String key, final String value)
+            throws IOException {
+        final File metadataFile = resolveMetadataFile(layerName);
+        Properties metadata = loadLayerMetadata(metadataFile);
+        // Current last modification time
+        final long lastModDate = metadataFile.lastModified();
+
+        boolean doUpdate;
+        String encodedValue;
+        if (null == value) {
+            doUpdate = metadata.containsKey(key);
+            encodedValue = null;
+        } else {
+            encodedValue = URLEncoder.encode(value, "UTF-8");
+            doUpdate = !Objects.equals(encodedValue, metadata.getProperty(key));
+        }
+        if (doUpdate) {
+            writeMetadataOptimisticLock(key, encodedValue, metadataFile, metadata, lastModDate);
+        }
+    }
+
+    private File getLayerPath(String layerName) {
+        String prefix = path + File.separator + filteredLayerName(layerName);
+        File layerPath = new File(prefix);
+        return layerPath;
+    }
+
+    private static String urlDecUtf8(String value) {
+        try {
+            value = URLDecoder.decode(value, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+        return value;
+    }
+
+    private int resolveLockBucket(File file) {
+        long consistentFileNameHash =
+                Hashing.farmHashFingerprint64()
+                        .hashString(file.getAbsolutePath(), StandardCharsets.UTF_8)
+                        .asLong();
+        int bucket = Hashing.consistentHash(consistentFileNameHash, locks.length);
+        return bucket;
+    }
+
+    private ReadWriteLock getLock(File file) {
+        return locks[resolveLockBucket(file)];
+    }
+
+    /**
+     * Performs the actual update of the metatada, making sure only the provided key/value pair is
+     * updated in case another process modified the metadata file since it was loaded by the caller
+     * code
+     *
+     * @throws IOException
+     */
+    @SuppressFBWarnings(value = "DLS_DEAD_LOCAL_STORE")
+    private void writeMetadataOptimisticLock(
+            final String key,
+            final String value,
+            final File metadataFile,
+            Properties metadata,
+            long lastModified)
+            throws IOException {
+        final ReadWriteLock rwLock = getLock(metadataFile);
+        final int maxAttempts = LayerMetadataStore.METADATA_MAX_RW_ATTEMPTS;
+
+        log.debug("Start attempt to add key (key: " + key + ")");
+
+        rwLock.writeLock().lock();
+        try {
+            createParentIfNeeded(metadataFile);
+            int attempt = 0;
+            for (attempt = 0; attempt < maxAttempts; attempt++) {
+                if (lastModified == metadataFile.lastModified()) {
+                    metadata = loadLayerMetadata(metadataFile);
```

Comment:
Since we rewrite the metadata object reference here, what's the point of passing them as a parameter? I don't see the value being used before.